src/coreclr/jit/fgdiagnostic.cpp

```diff
@@ -424,6 +424,51 @@ void Compiler::fgDumpTree(FILE* fgxFile, GenTree* const tree)
     }
 }
 
+namespace
+{
+    template<typename T>
+    class release_t final
+    {
+        T* _mem;
+    public:
+        release_t() : _mem{ nullptr }
+        { }
+
+        ~release_t()
+        {
+            reset();
+        }
+
+        void set(T* mem)
+        {
+            reset();
+            _mem = mem;
+        }
+
+        void reset()
+        {
+            free(_mem);
+            _mem = nullptr;
+        }
+    };
+
+    using CharRelease = release_t<char>;
+
+    char* NarrowString(LPCWSTR wideString, CharRelease& mem)
+    {
+        size_t len = wcslen(wideString) + 1;
+        char* alloc = (char*)malloc(sizeof(char) * len);
+        mem.set(alloc);
+        for (size_t i = 0; i < len; ++i)
+        {
+            assert(((unsigned)wideString[i]) <= 0x7f);
+            alloc[i] = (char)wideString[i];
+        }
```

Comment:
Can we call `WszWideCharToMultiByte` here to make it work for people with non-ASCII characters in their file paths? Or will the UTF8 in file paths not work?

At some point it will be nice to make `getStringConfigValue` return UTF8 (but not suggesting it needs to be done as part of this). We have other places in the JIT where we end up converting to UTF8 (`MethodSet::initialize`).

Reply:
> Can we call WszWideCharToMultiByte here to make it work for people with non-ASCII characters in their file paths? Or will the UTF8 in file paths not work?

I can convert the `WCHAR` to UTF8 and then back to `WCHAR`, but that transform is lossy and would likely just create more confusion when it fails. Is this a scenario that you or someone you know uses?

> At some point it will be nice to make getStringConfigValue return UTF8 (but not suggesting it needs to be done as part of this). We have other places in the JIT where we end up converting to UTF8 (MethodSet::initialize).

Agree.

Reply:
I don't personally have non-ASCII file paths, but I could imagine the LoongArch folks might.

If we call `fopen` with UTF8 path does that not work on Linux/Windows (with UTF8 enabled)?

Reply:
> I don't personally have non-ASCII file paths, but I could imagine the LoongArch folks might.

Yuck. You're right that isn't fair to them. Let me take your suggestion.